### ISSUE TITLE
External transaction queue

### ISSUE THREAD

**Issue by afck** [poanetwork/hbbft#104]
In many use cases, the user might want to implement their own transaction queue to e.g. prioritize messages. E.g. Parity already has a transaction queue.

We could make it a `Queue` trait, with the default implementation the current (internal) one. The user could then write their own implementation. A `Queue` would basically have to "return a random N-th of the first `batch_size` transactions" on demand, and allow removing a set of transactions after it has been output in a batch.

`HoneyBadger` would probably not have an `input` method at all then: the input would go directly to the queue?

**Comment by afck** [comment]
Another option is to just leave it to the user to call `input` exactly once per epoch, i.e.: Whenever Honey Badger outputs a new batch, the user should select the next proposal. Honey Badger itself would not have any queue at all, and "selecting one N-th of the first `batch_size` transactions" would turn into a recommendation to the user. It would also be the user's responsibility to remove output transactions from the queue. (We could still provide a convenient standard queue implementation, of course.)

Maybe in practice a Honey Badger Construction Kit is just more useful than a Honey Badger implementation itself?

**Comment by afck** [comment]
That would also make https://github.com/poanetwork/hbbft/issues/80 obsolete, as it would give the user full control over prioritization.

The `batch_size` parameter would probably be replaced by something limiting the size (in bytes) of other nodes' encrypted proposals.

For the default behavior, we could provide a completely separate queue implementation. After each batch, the user would then call something like:

```rust
// Remove the new batch's confirmed transactions from the queue.
queue.remove_transactions(batch.iter());
// Take 1/N of the first `batch_size` pending transactions and input them
// as the next proposal.
honey_badger.input(queue.create_proposal())?;
```

**Comment by vkomenda** [comment]
The second option that leaves `input` to the client of the library is more flexible in my opinion because it allows to use an existing queue in the client without making it an instance of a `Queue` trait. The client code would have no formal guidelines how to provide `input` though that would have been available if the library were to provide a trait API. This is not critical because we can document the `input` requirement and provide examples of integration into client code.

**Comment by afck** [comment]
Another complication is that the proposal size, which I'd like to [enforce by rejecting oversized proposals before even decrypting them](https://github.com/poanetwork/hbbft/issues/88), in `DynamicHoneyBadger` depends on both the number of validators and the particular batch size mode (fixed batch size, fixed proposal size, or _λ N² log(N)_).

I think the most flexible solution is:
* `HoneyBadger` has a configuration option `proposal_size`, which specifies the maximum in bytes.
* `Queue::generate_proposal` takes the `proposal_size` as an argument, and `Queue::new` takes a `max_transaction_size` so that it knows how many transactions it can fit into how many bytes.
* The `BatchSize` enum has variants to specify the mode. It helps compute the parameter for `HoneyBadger`.
* `DynamicHoneyBadger` takes a `BatchSize` as a configuration option, since it needs to dynamically recompute the proposal size in bytes whenever the validator set changes.

So the API will look something like this:

```rust
let mut dhb = DynamicHoneyBadger::new(netinfo)
    .with_batch_size(BatchSize::Fixed(1_000_000)) // 1 MB blocks
    .build()?;
let mut queue = Queue::new(1_000); // 1 kB max tx size
```

And then, whenever we received at least one new batch as output:

```rust
let mut had_output = false;
while let Some(batch) = dhb.next_output() {
    // Remove the confirmed transactions.
    queue.remove_transactions(batch.iter());
    had_output = true;
    do_application_level_logic(batch);
}
if had_output {
    // Get 1000 / N transactions as a proposal.
    let proposal = queue.generate_proposal(dhb.current_batch_size());
    // Input it for the next epoch.
    dhb.input(proposal)?;
}
```

**Comment by afck** [comment]
Implemented in #117.
